### ISSUE TITLE
OAuth2.AcquireTokenWithClientCredentials returning expired token after one hour.

### ISSUE THREAD

**Issue by KuljitSinghParmar** [microsoft/ALAppExtensions#15860]
**1. Describe the bug**
We are using procedure  OAuth2.AcquireTokenWithClientCredentials to acquire Access token . it works fine for one hour but after that it starts giving an  error - "Error Code: 401 , Error Message: The credentials provided are incorrect "  . When I get this error then if I restart my docker container, this error goes away. Not sure what's happening here .   if i use Postman to acquire the token , it works fine every time. See this post too. https://www.yammer.com/dynamicsnavdev/#/Threads/show?threadId=1487161678880768  

**2. To Reproduce**

Add a new button on any page and test there. 
``` AL
AccessToken is global variable.  

[NonDebuggable]
local procedure BssiGetAccessToken(BssiClientID: Text; BssiClientSecret: Text; BssiScope: Text; BssiAuthURL: Text; BssiTenantID: Text)
    var
        OAuth2: Codeunit OAuth2;
        BssiFailedAccessTokenErr: Label 'Failed to retrieve access token\';
        Scopes: List of [Text];
        OauthError: Text;
    begin
        Scopes.Add(BssiScope);
        clear(OAuth2);
        if not OAuth2.AcquireTokenWithClientCredentials(BssiClientID, BssiClientSecret,
            BssiGetAuthorityUrl(BssiAuthURL, BssiTenantID), '', Scopes, AccessToken) then
            Error(BssiFailedAccessTokenErr, GetLastErrorText());
    end;
	
 
local procedure GetAuthenticationHeaderValue(AccessToken: Text) Value: Text;
 begin
 Value := StrSubstNo('Bearer %1', AccessToken);
 end;

 
local procedure GetBCCompanyAPIUrl(TenantID: Text; BCEnvironmentName: Text) Url: Text;
 var
        TypeHelper: Codeunit "Type Helper";
 begin
        Url := GetCompanyURL; //BC URL 
        Url := Url.Replace('{AadTenantId}', TenantID)
                  .Replace('{BCEnvironmentName}', BCEnvironmentName);
 end;

	
Then i am using that access token to call an api -
	
   procedure CallBusinessCentralAPI(Resource: Text) Result: Text
    var
        BssiICPartner: Record BssiICPartnerSetup;
        BssiICEndpoints: Record BssiICEndpoints;
        BssiBlockedErr: Label 'Request was blocked by environment';
        BssiRequestFailedErr: Label 'Request to connect to IC Partner failed. Error Code: %1  , Error Message: %2', Comment = '%1 = Error Code;  %2=Error Phrase';
        Client: HttpClient;
        Response: HttpResponseMessage;
        Url: Text;
        Error: Text;
        ErrorMessage: Text;
        ErrorToShow: Text;
    begin
    
        BssiICEndpoints.Get(BssiICPartner.BssiEndpoint);
        BssiGetAccessToken(BssiICEndpoints.BssiClientID, BssiICEndpoints.BssiClientSecret,
                    BssiICEndpoints.BssiScopeUrl, BssiICEndpoints.BssiAuthUrl, BssiICEndpoints.BssiTenantID);

        Client.DefaultRequestHeaders.Add('Authorization', GetAuthenticationHeaderValue(AccessToken));
        Client.DefaultRequestHeaders.Add('Accept', 'application/json');
        url := GetBCCompanyAPIUrl(BssiICEndpoints.BssiTenantID, BssiICEndpoints.BssiEnvironment); 
        if not Client.Get(Url, Response) then
            if Response.IsBlockedByEnvironment then
                Error(BssiBlockedErr)
            else
                Error(BssiRequestFailedErr, GetLastErrorText());

        if not Response.IsSuccessStatusCode then 
            Error(BssiRequestFailedErr, Response.HttpStatusCode, Response.ReasonPhrase);

        Response.Content.ReadAs(Result);
    end;
	
	local procedure GetBCCompanyAPIUrl(TenantID: Text; BCEnvironmentName: Text) Url: Text;
    var
        TypeHelper: Codeunit "Type Helper";
    begin
        Url := GetCompanyURL;  //BC URL
        Url := Url.Replace('{AadTenantId}', TenantID)
                  .Replace('{BCEnvironmentName}', BCEnvironmentName);
    end;
```

**Note:** Because the developers need to copy and paste the code snippet, including a code snippet as a media file (i.e. .gif) is not sufficient.

**3. Expected behavior**
OAuth2.AcquireTokenWithClientCredentials  should return correct token every time it is used. 

**4. Actual behavior**
OAuth2.AcquireTokenWithClientCredentials  return correct token only for 1 hour afterwards it start giving expired tokens. 

**5. Versions:**

- AL Language: 8.2
- Visual Studio Code: 1.63.2
- Business Central: US Business Central 19.1 (Platform 19.0.31889.33801 + Application 19.1.31886.33837)
- List of Visual Studio Code extensions that you have installed:

### Final Checklist

Please remember to do the following:

* [ ] Search the issue repository to ensure you are reporting a new issue

* [ ] Reproduce the issue after disabling all extensions except the AL Language extension

* [ ] Simplify your code around the issue to better isolate the problem

**Comment by KuljitSinghParmar** [comment]
@AlexanderYakunin please review this.

**Comment by JesperSchulz** [comment]
Thanks for reporting this. We agree, and we’ll publish a fix asap, either in an update for the current version or in the next major release. Please do not reply to this, as we do not monitor closed issues. If you have follow-up questions or requests, please create a new issue where you reference this one.
